# Data cleaning and merging dataframes

## Loading multiple datasets

### Classical way

Reading file by file

In [1]:
import pandas as pd

import warnings
warnings.filterwarnings("ignore")

In [2]:
# path = '../data/eniac/'
path = '../eniac_dataset/'
orderlines = pd.read_csv(path + 'orderlines.csv')
orders = pd.read_csv(path + 'orders.csv')
brands = pd.read_csv(path + 'brands.csv')
products = pd.read_csv(path + 'products.csv')

df_list = [orderlines, orders, brands, products]
files = ['orderlines','orders','brands','products']

### Another way

Using `os` and a loop to read all the files from a directory. It's also possible to read only files with a certain extension (like `.csv`):

In [3]:
# import pandas as pd
# import os
# path = '../data/eniac/'
# path, dirs, files = next(os.walk(path))
# #print(files)

# # remove non-csv files
# for file in files:
#     if not file.endswith("csv"):
#         files.remove(file)

In [4]:
# # create empty list 
# df_list = []

# # append datasets to the list
# for file in files:
#     temp_df = pd.read_csv(path + file, sep=',')
#     df_list.append(temp_df)

# # show results
# products, orders, orderlines, brands = df_list[0], df_list[1], df_list[2], df_list[3]

## Data quality

### Missing values

In [5]:
def check_missing_values(data): 
    print('Missing values:' + '\n' + str(data.isna().sum()))

In [6]:
for i in range(len(df_list)): 
    print(f'File name: {files[i]}')
    check_missing_values(df_list[i])
    print('\n')

File name: orderlines
Missing values:
id                  0
id_order            0
product_id          0
product_quantity    0
sku                 0
unit_price          0
date                0
dtype: int64


File name: orders
Missing values:
order_id        0
created_date    0
total_paid      5
state           0
dtype: int64


File name: brands
Missing values:
short    0
long     0
dtype: int64


File name: products
Missing values:
sku             0
name            0
desc            7
price          46
promo_price     0
in_stock        0
type           50
dtype: int64




### Duplicates

In [7]:
def check_duplicates(data): 
    print('Duplicated rows: ', data.duplicated().sum())

In [8]:
for i in range(len(df_list)): 
    print(f'File name: {files[i]}')
    check_missing_values(df_list[i])
    check_duplicates(df_list[i])
    print('\n')

File name: orderlines
Missing values:
id                  0
id_order            0
product_id          0
product_quantity    0
sku                 0
unit_price          0
date                0
dtype: int64
Duplicated rows:  0


File name: orders
Missing values:
order_id        0
created_date    0
total_paid      5
state           0
dtype: int64
Duplicated rows:  0


File name: brands
Missing values:
short    0
long     0
dtype: int64
Duplicated rows:  0


File name: products
Missing values:
sku             0
name            0
desc            7
price          46
promo_price     0
in_stock        0
type           50
dtype: int64
Duplicated rows:  8746




Based on our initial exploration we need to find a way to deal with our missing values. 

Do we have other problems like data inconsistency? 

* Example: problems with the `price` column in the `orderlines` dataframe.

### Data inconsistency

In [9]:
for i in range(len(df_list)): 
    
    print(f'File name: {files[i]}')
    check_missing_values(df_list[i])
    check_duplicates(df_list[i])
    df_list[i].info()
    print('\n')

File name: orderlines
Missing values:
id                  0
id_order            0
product_id          0
product_quantity    0
sku                 0
unit_price          0
date                0
dtype: int64
Duplicated rows:  0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


File name: orders
Missing values:
order_id        0
created_date    0
total_paid      5
state           0
dtype: int64
Duplicated rows:  0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 

Some aspects that do not make a lot of sense:

* **products**: 
    * `price` and `promo_price` are loaded as objects. They should be floats. 
    * Quite a lot of missing values: 
        * `description`: 7 missing values. Maybe that could be inferred from the product name?
        * `price`: the missing values could be filled from the `orderlines` dataset. But first we will need to clean it. 

    
* **orders**: 
    * `created_date` should have a date data type. Then, it would be a good quality check to see if the created date for `orders` mathces with the created dates for `orderlines`. 
    
    
* **orderlines**: 
    * `unit_price` has to be a float, something wrong there. 
    * `date` has to be transformed to a date data type. Then, as we said, check that it with matches with the `orders` dataset. 
    
    
* **brands**: looks fine. 


Where do we have to start? Since `orders` and `orderlines` seem very crucial to the analysis, we will start by checking that they match.

## Cleaning orders

In [10]:
# change date datatype
orders['created_date'] = pd.to_datetime(orders['created_date'])

Some data I could extract from here to do a data quality check with `orderlines`: 

* the sum of `total_paid` on both datasets should be the same
* the initial and last date should be the same

In [11]:
min(orders['created_date'])

Timestamp('2017-01-01 00:07:19')

In [12]:
max(orders['created_date'])

Timestamp('2018-03-14 13:58:36')

In [13]:
sum(orders['total_paid'])

nan

I forgot to deal with missing values on our orders dataset!

In [14]:
orders.isna().sum()

order_id        0
created_date    0
total_paid      5
state           0
dtype: int64

In [15]:
# let's filter them out
orders.loc[orders['total_paid'].isna()]

order_id        created_date  total_paid    state
127701    427314 2017-11-20 18:54:39         NaN  Pending
132013    431655 2017-11-22 12:15:24         NaN  Pending
147316    447411 2017-11-27 10:32:37         NaN  Pending
148833    448966 2017-11-27 18:54:15         NaN  Pending
149434    449596 2017-11-27 21:52:08         NaN  Pending

These orders have a state of "Pending", so I can exclude them from our dataset. We are interested on the completed ones.

In [16]:
orders.dropna(inplace=True)

Now my dataset is clean. And the total paid is: 

In [17]:
orders['total_paid'].sum()

129159615.07000001

## Cleaning orderlines

First let's transform our date time. 

In [18]:
orderlines['date'] = pd.to_datetime(orderlines['date'])

In [19]:
min(orderlines['date'])
# orders: Timestamp('2017-01-01 00:07:19')

Timestamp('2017-01-01 00:07:19')

In [20]:
max(orderlines['date'])
# orders: Timestamp('2018-03-14 13:58:36')

Timestamp('2018-03-14 13:58:36')

Yes, it seems it match. Now let's test our total_paid. 

To get this value, we will have to calculate a new column, total price for each row. It would be `product_quantity` * `unit_price`. I can not multiply a two different types of data. So I will have to transform the unit price to a numeric one. 

In [21]:
# orderlines['product_quantity'] * pd.to_numeric(orderlines['unit_price'])

While trying to transform this column to numeric an error appears. From the error message, "Unable to parse string "1.137.99" at position 6", it seems that our dataset has some problems with the thousands separators: they were encoded as dots, and python only admits one dot per number: the _decimal_ separator!

There are a lot of different ways to solve this problem. The first thing I will do is to count how many dots appear for each number using string operations. If price has more than one dot, I will consider it corrupted. 

In [22]:
# we create a copy of the dataset
ol_temp = orderlines.copy()

# create a new column with the amount of dots in the unit_price column
ol_temp['dots'] = orderlines['unit_price'].str.count('\.')

# show the rows with more than one dot
ol_temp.query('dots > 1')

id  id_order  product_id  product_quantity      sku unit_price  \
6       1119115    299544           0                 1  APP1582   1.137.99   
11      1119126    299549           0                 1  PAC0929   2.565.99   
15      1119131    299553           0                 1  APP1854   3.278.99   
43      1119195    299582           0                 1  PAC0961   2.616.99   
59      1119214    299596           0                 1  PAC1599   2.873.99   
...         ...       ...         ...               ...      ...        ...   
293862  1649999    452946           0                 1  APP2075   2.999.00   
293887  1650045    527321           0                 1  PAC2148   3.497.00   
293889  1650050    527324           0                 1  PAC2117   3.075.00   
293911  1650088    527342           0                 1  APP2492   1.329.00   
293936  1650135    527364           0                 1  APP2491   1.329.00   

                      date  dots  
6      2017-01-01 01:17:21     2  
11     2017-01-01 02:07:42     2  
15     2017-01-01 02:14:47     2  
43     2017-01-01 08:54:00     2  
59     2017-01-01 09:53:11     2  
...                    ...   ...  
293862 2018-03-14 13:03:33     2  
293887 2018-03-14 13:10:15     2  
293889 2018-03-14 13:10:56     2  
293911 2018-03-14 13:24:51     2  
293936 2018-03-14 13:36:13     2  

[36169 rows x 8 columns]

Our theory about the thousands separators is confirmed. How can we solve this problem? 

My approach would be to remove all the dots for all the unit_prices, and then add a dot before the last 2 digits to all the rows. Then I will transform it into numeric values.

In [23]:
orderlines = orderlines.assign(unit_price_nd = orderlines['unit_price'].str.replace('\.','', regex=True))
orderlines.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                 date unit_price_nd  
0 2017-01-01 00:07:19          1899  
1 2017-01-01 00:19:45         39900  
2 2017-01-01 00:20:57         47405  
3 2017-01-01 00:51:40          6839  
4 2017-01-01 01:06:38          2374

In [24]:
orderlines['integers'] = orderlines['unit_price_nd'].str[:-2]
orderlines['digits'] = orderlines['unit_price_nd'].str[-2:]

orderlines.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                 date unit_price_nd integers digits  
0 2017-01-01 00:07:19          1899       18     99  
1 2017-01-01 00:19:45         39900      399     00  
2 2017-01-01 00:20:57         47405      474     05  
3 2017-01-01 00:51:40          6839       68     39  
4 2017-01-01 01:06:38          2374       23     74

In [25]:
orderlines['new_unit_price'] = orderlines['integers'] + '.' + orderlines['digits']
orderlines.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                 date unit_price_nd integers digits new_unit_price  
0 2017-01-01 00:07:19          1899       18     99          18.99  
1 2017-01-01 00:19:45         39900      399     00         399.00  
2 2017-01-01 00:20:57         47405      474     05         474.05  
3 2017-01-01 00:51:40          6839       68     39          68.39  
4 2017-01-01 01:06:38          2374       23     74          23.74

In [26]:
# last step is to transform it to numeric
orderlines['unit_price'] = pd.to_numeric(orderlines['new_unit_price'])
orderlines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        293983 non-null  float64       
 6   date              293983 non-null  datetime64[ns]
 7   unit_price_nd     293983 non-null  object        
 8   integers          293983 non-null  object        
 9   digits            293983 non-null  object        
 10  new_unit_price    293983 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(4), object(5)
memory usage: 24.7+ MB


Now it is time to see if it is the same than the orders files. 

In [27]:
orderlines.drop(['unit_price_nd','digits','integers','new_unit_price'], axis=1, inplace=True)

In [28]:
orderlines.head()

id  id_order  product_id  product_quantity      sku  unit_price  \
0  1119109    299539           0                 1  OTT0133       18.99   
1  1119110    299540           0                 1  LGE0043      399.00   
2  1119111    299541           0                 1  PAR0071      474.05   
3  1119112    299542           0                 1  WDT0315       68.39   
4  1119113    299543           0                 1  JBL0104       23.74   

                 date  
0 2017-01-01 00:07:19  
1 2017-01-01 00:19:45  
2 2017-01-01 00:20:57  
3 2017-01-01 00:51:40  
4 2017-01-01 01:06:38

In [29]:
orderlines['total_price'] = orderlines['product_quantity'] * orderlines['unit_price']
sum(orderlines['total_price'])
# 129159615.07000001

128776222.0298095

In [30]:
orderlines.head()

id  id_order  product_id  product_quantity      sku  unit_price  \
0  1119109    299539           0                 1  OTT0133       18.99   
1  1119110    299540           0                 1  LGE0043      399.00   
2  1119111    299541           0                 1  PAR0071      474.05   
3  1119112    299542           0                 1  WDT0315       68.39   
4  1119113    299543           0                 1  JBL0104       23.74   

                 date  total_price  
0 2017-01-01 00:07:19        18.99  
1 2017-01-01 00:19:45       399.00  
2 2017-01-01 00:20:57       474.05  
3 2017-01-01 00:51:40        68.39  
4 2017-01-01 01:06:38        23.74

In [31]:
sum(orders['total_paid'])

129159615.06991275

What!? After all this work the values do not match! How can we figure it out where is the problem? 

## Matching `orders` and `orderlines`

It is possible that some orders exist in one dataset but not in the other one. This would be a potential source for this price mismatch. Let's solve it!

We first create a new column in the `orderlines` dataset using `assign`. We also use `isin()` to create a boolen value (True/False) that checks whether the `id_order` is present in the `orders` dataset:

In [32]:
orderlines.assign(check_orders = orderlines['id_order'].isin(orders['order_id']))

id  id_order  product_id  product_quantity      sku  unit_price  \
0       1119109    299539           0                 1  OTT0133       18.99   
1       1119110    299540           0                 1  LGE0043      399.00   
2       1119111    299541           0                 1  PAR0071      474.05   
3       1119112    299542           0                 1  WDT0315       68.39   
4       1119113    299543           0                 1  JBL0104       23.74   
...         ...       ...         ...               ...      ...         ...   
293978  1650199    527398           0                 1  JBL0122       42.99   
293979  1650200    527399           0                 1  PAC0653      141.58   
293980  1650201    527400           0                 2  APP0698        9.99   
293981  1650202    527388           0                 1  BEZ0204       19.99   
293982  1650203    527401           0                 1  APP0927       13.99   

                      date  total_price  check_orders  
0      2017-01-01 00:07:19        18.99          True  
1      2017-01-01 00:19:45       399.00          True  
2      2017-01-01 00:20:57       474.05          True  
3      2017-01-01 00:51:40        68.39          True  
4      2017-01-01 01:06:38        23.74          True  
...                    ...          ...           ...  
293978 2018-03-14 13:57:25        42.99          True  
293979 2018-03-14 13:57:34       141.58          True  
293980 2018-03-14 13:57:41        19.98          True  
293981 2018-03-14 13:58:01        19.99          True  
293982 2018-03-14 13:58:36        13.99          True  

[293983 rows x 9 columns]

Then, using `.query` we select rows where the value in this new column is `False`:

In [33]:
(
orderlines
    .assign(check_orders = orderlines['id_order'].isin(orders['order_id']))
    
)['check_orders'].sum()

293743

In [34]:
orderlines.shape

(293983, 8)

In [35]:
(
orderlines
    .assign(check_orders = orderlines['id_order'].isin(orders['order_id']))
    .query("check_orders==False")
)

id  id_order  product_id  product_quantity      sku  unit_price  \
5       1119114    295310           0                10  WDT0249      231.79   
63      1119218    296284           0                 1  BNQ0042      699.00   
67      1119223    294806           0                 1  APP1849     2558.99   
69      1119226    294806           0                 1  APP1864     2797.99   
70      1119235    297261           0                 1  QNA0177      304.99   
...         ...       ...         ...               ...      ...         ...   
275665  1621177    244328           0                 1  OWC0260      349.00   
280055  1628699    261391           0                 1  APP2352     3343.00   
280352  1629247    287797           0                 1  SYN0182      484.11   
280856  1630150    261391           0                 1  SAT0091       44.99   
280869  1630178    261391           0                 1  SAT0062       49.99   

                      date  total_price  check_orders  
5      2017-01-01 01:14:27      2317.90         False  
63     2017-01-01 09:58:35       699.00         False  
67     2017-01-01 10:09:15      2558.99         False  
69     2017-01-01 10:15:14      2797.99         False  
70     2017-01-01 10:17:59       304.99         False  
...                    ...          ...           ...  
275665 2018-02-20 13:00:45       349.00         False  
280055 2018-02-27 11:43:25      3343.00         False  
280352 2018-02-27 19:16:26       484.11         False  
280856 2018-02-28 12:18:26        44.99         False  
280869 2018-02-28 12:30:57        49.99         False  

[240 rows x 9 columns]

In [36]:
(
orderlines
    .assign(check_orders = orderlines['id_order'].isin(orders['order_id']))
    .query("check_orders==False")
).shape

(240, 9)

It looks like 240 rows in `orderlines` come from orders not present in the `orders` dataset. This is quite inconsistent, since the `orders` dataset should be the one and only source of truth for orders: if an order is not there, it should not exist. We will try to find out why that happened, but for now, let's just remove those "ghost" orders:

In [37]:
orderlines.shape

(293983, 8)

In [38]:
# drop the order id that does not present in orders
orderlines = (
orderlines
    .assign(check_orders = orderlines['id_order'].isin(orders['order_id']))
    .query("check_orders==True")
             )
orderlines

id  id_order  product_id  product_quantity      sku  unit_price  \
0       1119109    299539           0                 1  OTT0133       18.99   
1       1119110    299540           0                 1  LGE0043      399.00   
2       1119111    299541           0                 1  PAR0071      474.05   
3       1119112    299542           0                 1  WDT0315       68.39   
4       1119113    299543           0                 1  JBL0104       23.74   
...         ...       ...         ...               ...      ...         ...   
293978  1650199    527398           0                 1  JBL0122       42.99   
293979  1650200    527399           0                 1  PAC0653      141.58   
293980  1650201    527400           0                 2  APP0698        9.99   
293981  1650202    527388           0                 1  BEZ0204       19.99   
293982  1650203    527401           0                 1  APP0927       13.99   

                      date  total_price  check_orders  
0      2017-01-01 00:07:19        18.99          True  
1      2017-01-01 00:19:45       399.00          True  
2      2017-01-01 00:20:57       474.05          True  
3      2017-01-01 00:51:40        68.39          True  
4      2017-01-01 01:06:38        23.74          True  
...                    ...          ...           ...  
293978 2018-03-14 13:57:25        42.99          True  
293979 2018-03-14 13:57:34       141.58          True  
293980 2018-03-14 13:57:41        19.98          True  
293981 2018-03-14 13:58:01        19.99          True  
293982 2018-03-14 13:58:36        13.99          True  

[293743 rows x 9 columns]

In [39]:
orderlines.shape

(293743, 9)

Now let's look at this problem in the opposite direction: are there orders in the `orders` dataset not prsent in `orderlines`?

In [40]:
# look the other way round
(
    orders
 .assign(check_orders = orders['order_id'].isin(orderlines['id_order']))
 .query("check_orders==False")
)

order_id        created_date  total_paid            state  \
8         245941 2017-01-01 10:32:23      183.52        Completed   
65        268629 2017-01-31 11:27:25       73.23        Completed   
75        272862 2017-04-18 18:17:10      128.99          Pending   
82        277655 2017-02-01 08:44:55        9.99          Pending   
83        277994 2017-01-23 18:30:11       52.99        Completed   
...          ...                 ...         ...              ...   
226835    527328 2018-03-14 13:14:12        0.00  Shopping Basket   
226851    527344 2018-03-14 13:25:21        0.00      Place Order   
226853    527346 2018-03-14 13:26:14        0.00      Place Order   
226855    527348 2018-03-14 13:28:18        0.00      Place Order   
226889    527382 2018-03-14 13:47:43        0.00  Shopping Basket   

        check_orders  
8              False  
65             False  
75             False  
82             False  
83             False  
...              ...  
226835         False  
226851         False  
226853         False  
226855         False  
226889         False  

[22213 rows x 5 columns]

There are more than 22000 orders in the `orders` dataset that are not present on the `orderlines` dataset!!! We can try to find out why by looking at the state of these orders:

In [41]:
(orders
 .assign(check_orders = orders['order_id'].isin(orderlines['id_order']))
 .query("check_orders==False")
 ['state'].value_counts())

Place Order        12304
Shopping Basket     9810
Completed             45
Cancelled             41
Pending               13
Name: state, dtype: int64

It looks like most of them are orders that were not fully completed: the products were left in the shopping basket or the order was "placed" but maybe not paid (hence the state "Place Order". Some of them were "Completed", though. 

This will require further research, and we might have to come back to these orders if we have to explore consumer behaviour (e.g. why are orders left in the shopping basket?), but for now, for the sake of data quality, let's drop all of these unmatched orders:

In [42]:
orders = (orders
          .assign(check_orders = orders['order_id'].isin(orderlines['id_order']))
          .query("check_orders==True")
         )

Let's now check again if the total paid matches:

In [43]:
orders['total_paid'].sum()

129130877.18

In [44]:
orderlines['total_price'].sum()

128659393.77

In [45]:
orders['order_id'].count()#.columns

204691

STILL NOT MATCHING!!! This is outrageous. Let's keep exploring.

In [46]:
orderlines['id_order'].count()

293743

## Solving the price mismatch

Let's merge both datasets and compare, order by order, the `total_price`. We will call this new merged dataset `orders_info`.

*Note: To merge `orderlines` with `orders`, we have to group and aggregate `orderlines`!

In [47]:
orders_info = (
orderlines
    .groupby('id_order')
    .agg({'total_price':'sum'})
    .merge(orders, how='left', left_on='id_order', right_on='order_id')
    .copy()
)
orders_info

total_price  order_id        created_date  total_paid  \
0             44.99    241319 2017-01-02 13:35:40       44.99   
1            129.16    241423 2017-11-06 13:10:02      136.15   
2             10.77    242832 2017-12-31 17:40:03       15.76   
3             77.99    243330 2017-02-16 10:59:38       84.98   
4            153.87    243784 2017-11-24 13:35:19      157.86   
...             ...       ...                 ...         ...   
204686        42.99    527397 2018-03-14 13:56:38       42.99   
204687        42.99    527398 2018-03-14 13:57:25       42.99   
204688       141.58    527399 2018-03-14 13:57:34      141.58   
204689        19.98    527400 2018-03-14 13:57:41       19.98   
204690        13.99    527401 2018-03-14 13:58:36       18.98   

                  state  check_orders  
0             Cancelled          True  
1             Completed          True  
2             Completed          True  
3             Completed          True  
4             Cancelled          True  
...                 ...           ...  
204686      Place Order          True  
204687  Shopping Basket          True  
204688  Shopping Basket          True  
204689  Shopping Basket          True  
204690      Place Order          True  

[204691 rows x 6 columns]

Now that the `total` from both datasets is in the same dataframe, we can create a new column with the difference:

In [48]:
orders_info['price_difference'] = orders_info['total_price'] - orders_info['total_paid']
orders_info.sort_values('price_difference').tail(30)

total_price  order_id        created_date  total_paid  \
20654        164.90    322147 2017-02-10 11:56:59      125.99   
11725        164.90    312311 2017-01-20 00:59:54      125.99   
45940         44.99    349377 2017-04-24 21:42:28        4.99   
8545         164.90    308825 2017-01-13 14:27:46      123.99   
21347        164.90    322901 2017-02-12 13:25:12      122.99   
25517        164.90    327363 2017-02-21 22:42:17      122.99   
13784        164.90    314570 2017-01-25 09:01:35      122.99   
986          164.90    300474 2017-01-02 12:24:54      119.00   
166458    214768.02    484334 2018-01-07 15:58:26   214714.31   
54853      13450.87    359083 2017-05-25 13:58:31    13391.86   
134183       341.98    448286 2017-11-27 14:04:40      275.99   
18563        142.44    319901 2017-02-04 16:09:34       60.43   
14391        151.99    315229 2017-01-31 10:57:07       66.98   
148204        89.99    463567 2017-12-15 09:45:36        0.00   
111          135.97    289685 2017-01-09 23:15:07       15.95   
102967       441.12    412943 2017-10-16 09:51:14      290.54   
120052       426.86    432303 2017-11-22 21:43:02      261.86   
46945        339.99    350460 2017-04-27 18:44:21        6.99   
141740       359.99    456491 2017-12-06 20:43:59        0.00   
151599       770.98    467341 2017-12-19 23:10:24      408.98   
194349       388.06    515818 2018-02-22 01:02:35        6.59   
186388      1124.00    506820 2018-02-06 13:34:20        1.33   
179022      1558.00    498550 2018-01-25 17:57:06        3.81   
162308      1958.36    479621 2018-01-02 13:33:37       26.38   
201739      1968.00    524152 2018-03-09 21:00:48        6.59   
192913      2342.00    514198 2018-02-19 11:40:55        6.59   
193911      3038.00    515325 2018-02-21 11:31:20       13.19   
188169      3112.59    508818 2018-02-09 13:43:30        6.59   
197889      7771.23    519801 2018-03-02 08:59:34       31.22   
56804     250802.37    361200 2017-06-01 11:55:02   214642.47   

                  state  check_orders  price_difference  
20654         Completed          True             38.91  
11725         Completed          True             38.91  
45940           Pending          True             40.00  
8545          Completed          True             40.91  
21347       Place Order          True             41.91  
25517         Completed          True             41.91  
13784         Completed          True             41.91  
986           Completed          True             45.90  
166458  Shopping Basket          True             53.71  
54853   Shopping Basket          True             59.01  
134183      Place Order          True             65.99  
18563         Completed          True             82.01  
14391         Cancelled          True             85.01  
148204  Shopping Basket          True             89.99  
111           Completed          True            120.02  
102967  Shopping Basket          True            150.58  
120052        Completed          True            165.00  
46945         Cancelled          True            333.00  
141740  Shopping Basket          True            359.99  
151599  Shopping Basket          True            362.00  
194349  Shopping Basket          True            381.47  
186388  Shopping Basket          True           1122.67  
179022  Shopping Basket          True           1554.19  
162308  Shopping Basket          True           1931.98  
201739  Shopping Basket          True           1961.41  
192913  Shopping Basket          True           2335.41  
193911  Shopping Basket          True           3024.81  
188169  Shopping Basket          True           3106.00  
197889  Shopping Basket          True           7740.01  
56804   Shopping Basket          True          36159.90

In [49]:
orders_info['price_difference'].describe()

count    204691.000000
mean         -2.303391
std         297.461827
min      -90898.610000
25%          -0.010000
50%           0.000000
75%           0.000000
max       36159.900000
Name: price_difference, dtype: float64

Looks like the maximum and minimun price differences are huge: some orders are really corrupted. But we also see from the mean and the quartiles that the price difference is neglegible for most orders.

## Challenge: Remove outliers

Decide on a criteria for removing orders whenever you cannot trust the price difference between `orders` and `orderlines`. 

In [50]:
# code here

Save the data once you are sure you can trust it!

In [51]:
orderlines.to_csv(path + 'orderlines_cl.csv', index=False)
orders.to_csv(path + 'orders_cl.csv', index=False)

## Challenge: Cleaning products

Now it is time to clean the products dataset. Let's do a quick review of the things where wrong: 

In [52]:
print(products.info())
check_duplicates(products)
check_missing_values(products)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB
None
Duplicated rows:  8746
Missing values:
sku             0
name            0
desc            7
price          46
promo_price     0
in_stock        0
type           50
dtype: int64


Looking at this overview, we can see that there are different things that has to be changed: 

* Data types: 
    * price should be a float
    * promo price should be a float
* Duplicated rows. They have to be removed. 
    * To accomplish this step you can use the method `pd.DataFrame.drop_duplicates()`. Be sure you drop all the duplicates based on the column **sku**, as it is the one you will use to merge with orderlines. 
* Missing values: 
    * Description may be can be infered by the name
    * price. Is there a way we can extract the information from another table?
    * type. Do we need this column for our analysis?
    
This task can be accomoplished by using all the methods you already know.

### Start of the challenge

#### Duplicates

In [53]:
products.shape

(19326, 7)

In [54]:
products.drop_duplicates(inplace=True)

In [55]:
products.shape

(10580, 7)

In [56]:
products.head()

sku                                           name  \
0  RAI0007              Silver Rain Design mStand Support   
1  APP0023              Apple Mac Keyboard Keypad Spanish   
2  APP0025               Mighty Mouse Apple Mouse for Mac   
3  APP0072  Apple Dock to USB Cable iPhone and iPod white   
4  KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   

                                                desc  price promo_price  \
0       Aluminum support compatible with all MacBook  59.99     499.899   
1          USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                             mouse Apple USB cable.     59     569.898   
3              IPhone dock and USB Cable Apple iPod.     25     229.997   
4  2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   

   in_stock      type  
0         1      8696  
1         0  13855401  
2         0      1387  
3         0      1230  
4         1      1364

Check if you have a unique description for each sku: 

In [57]:
products.groupby('sku')['sku'].count().sort_values(ascending=False).head(5)

sku
APP1197    2
PAC1555    1
PAC1557    1
PAC1558    1
PAC1559    1
Name: sku, dtype: int64

The product APP1197 is not unique. Let's zoom in: 

In [58]:
products.query('sku == "APP1197"')

sku                                               name  \
7992  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   
8000  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   

                                                   desc price promo_price  \
7992  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...  1729     1305.59   
8000  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...   NaN     1305.59   

      in_stock  type  
7992         0  1282  
8000         0  1282

In [59]:
products.drop_duplicates('sku').query('sku == "APP1197"')

sku                                               name  \
7992  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   

                                                   desc price promo_price  \
7992  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...  1729     1305.59   

      in_stock  type  
7992         0  1282

I can drop the duplicates only focusing on the column 'sku': 

In [60]:
products.drop_duplicates('sku', inplace=True)

#### Data types

In [61]:
# pd.to_numeric(products['price'])

It seems this time that the problem in our products prices is different than the one we have found on orderlines. How is it possible we have more than 3 digits? 

Let's look at a random sample in our dataframe to understant more about it: 

In [62]:
products.sample(10)

sku                                               name  \
19079    APP2605                   Apple 42mm Sport Black Belt Loop   
18331    PHI0084  Philips Hue White and Color Ambiance Kit Incio...   
12000    BEL0197                        Belkin Car Charger White 1A   
15626    APP2004  Apple Watch Series 2 42mm Aluminum Gold Case a...   
14569    SAT0024          Satechi Aluminum Rose Gold iPhone Support   
18711  TUC0266-A         Open - Tucano iPad Case Giro Pro 97 "Black   
7037     PAC1612  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 32GB...   
15862    BEZ0178  Be.ez LArobe MacBook Pro Retina ONE sleeve 13 ...   
11897    QDO0009  QDOS OptiGuard Titanium full protective iPhone...   
2915   SAN0040-A   Open - Ultra SanDisk Ultra 64GB USB 3.0 PenDrive   

                                                    desc   price promo_price  \
19079  Durable and flexible strap sports lockable cli...      59     569.995   
18331  Philips Hue Starter Kit White and Color Ambian...  199.99   1.689.898   
12000   Car Charger 5W 1A USB port for universal iPhone.    9.99      74.996   
15626  Apple Watch Series 2 42 mm with sports strap i...     469   4.689.996   
14569  Stand with aluminum finish Lightning access to...   34.99     249.901   
18711          Reconditioned Rotary Case for iPad Pro 97   39.99     259.844   
7037   Desktop computer iMac 27-inch 3.3GHz Core i5 5...    4789  34.399.901   
15862   Neoprene Sleeve for MacBook Pro Retina 13 inches   29.95     249.865   
11897  Crystal screen protector with metal finishes i...   24.99     119.899   
2915   USB 3.0 flash drive 64GB USB Flash Drive Mac a...      31     176.946   

       in_stock      type  
19079         0      2449  
18331         1  11905404  
12000         0  13615399  
15626         0  24895185  
14569         1      5720  
18711         0  12635403  
7037          0  5,74E+15  
15862         0  13835403  
11897         0  13555403  
2915          0      1298

My strategy would be: 

1. Replace missing values with a "special" string. In that case I will put "000.000" as it would represent that this value has a wrong format and has to be changed. This step is done to avoid future problems while cleaning the column price.
2. Create a new column called "price_split" with the values of price splited by the `.`
3. Create a new column called "dots_count" counting the number of dots a price contains

In [63]:
import numpy as np

products = (
products
    # Replace all missing values by '000.000'
    .assign(price = lambda x: x['price'].replace(np.nan, '000.000'))
    .assign(price_split = lambda x: x['price'].str.split('\.'),
            dots_count = lambda x: x['price'].str.count('\.'))
)
products.head(10)

sku                                               name  \
0    RAI0007                  Silver Rain Design mStand Support   
1    APP0023                  Apple Mac Keyboard Keypad Spanish   
2    APP0025                   Mighty Mouse Apple Mouse for Mac   
3    APP0072      Apple Dock to USB Cable iPhone and iPod white   
4    KIN0007        Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
5    APP0073     Apple Composite AV Cable iPhone and iPod white   
6    KIN0008        Mac Memory Kingston 1GB 667MHz DDR2 SO-DIMM   
7    KIN0009        Mac Memory Kingston 2GB 800MHz DDR2 SO-DIMM   
8  KIN0001-2  Mac memory Kingston 4GB (2x2GB) 667MHz DDR2 SO...   
9    APP0100             Apple Adapter Mini Display Port to VGA   

                                                desc  price promo_price  \
0       Aluminum support compatible with all MacBook  59.99     499.899   
1          USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                             mouse Apple USB cable.     59     569.898   
3              IPhone dock and USB Cable Apple iPod.     25     229.997   
4  2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
5  IPhone and iPod AV Cable Dock to Composite Video.     45     420.003   
6  1GB RAM Mac mini and iMac (2006/07) MacBook Pr...  18.99     146.471   
7       2GB RAM iMac with Intel Core 2 Duo (Penryn).  36.99     274.694   
8  RAM 4GB (2x2GB) Mac mini and iMac (2006/07) Ma...     74     669.904   
9  Adapter Mini Display Port to VGA MacBook and M...     35     330.003   

   in_stock      type price_split  dots_count  
0         1      8696    [59, 99]           1  
1         0  13855401        [59]           0  
2         0      1387        [59]           0  
3         0      1230        [25]           0  
4         1      1364    [34, 99]           1  
5         0      1230        [45]           0  
6         0      1364    [18, 99]           1  
7         0      1364    [36, 99]           1  
8         0      1364        [74]           0  
9         0      1325        [35]           0

In [64]:
products.sample(10)

sku                                               name  \
12311    DLL0043       Dell P2717H Monitor 27 "IPS DP 1.2 pivotably   
1834     WAC0163  Wacom Cintiq Interactive Pen & Touch 27QHD Mac...   
11519  LOG0172-A  Open - Multi-Device Logitech K480 Keyboard for...   
13005    LAC0189  LaCie Porsche Design Hard Drive 5TB USB-C and ...   
15482  CRU0036-A  Open - Memory Crucial Mac 2GB 667MHZ DDR2 SO-DIMM   
15964    PAC1955  Synology DS1817 + | 2GB RAM | 16TB (8x2TB) Sea...   
2487   OWC0012-A  Open - OWC Mercury Aura Pro SSD 120GB MacBook ...   
2675     APP1045                                 Apple USB-C to USB   
12994    TPL0028  TP-Link TL-SG1024D Desktop Switch 24 Gigabit E...   
932      CRU0023        Crucial Mac Memory 4GB DDR3 1333MHz SO-DIMM   

                                                    desc    price promo_price  \
12311                          Pivotable Monitor for Mac   359.37   2.689.891   
1834   Professional graphics tablet touch screen with...   2549.9  22.139.975   
11519     Keying with bluetooth for iPad iPhone and Mac.    49.99     405.991   
13005  Desktop Hard Drive 5TB with USB-C adapter cabl...   219.99   1.977.939   
15482  2GB RAM Mac mini and iMac (2006/07) MacBook Pr...    35.99     274.606   
15964  NAS server with 2GB of RAM and 16TB for Mac an...  1858.81   1.574.746   
2487   120GB SSD drive with open MacBook Air 2008/200...   175.99   1.224.715   
2675                      USB-C USB adapter for MacBook.       25     240.004   
12994  Desktop Switch with 24 ports Gigabit RJ-45 10/...   120.88     829.903   
932    4GB RAM Mac mini iMac (2010/11) and MacBook Pr...    53.99     529.895   

       in_stock      type price_split  dots_count  
12311         0      1296   [359, 37]           1  
1834          0      1405   [2549, 9]           1  
11519         0      1298    [49, 99]           1  
13005         0  11935397   [219, 99]           1  
15482         0      1298    [35, 99]           1  
15964         0  12175397  [1858, 81]           1  
2487          0  12215397   [175, 99]           1  
2675          1  12585395        [25]           0  
12994         0      1334   [120, 88]           1  
932           1      1364    [53, 99]           1

Now I need to look at each value in the new column "price_split" and check the folliwing condition: 

If the last element of the value in price split has a length bigger than 2 then add a `True` inside the list `need_check`, else add a `False`. 

Example 1: 

 - Input: price_split = `[[59],[99]]` The last element of the value in price split is `[99]`. The length of this element is 2, as it only have 2 characters. Then we will append a `False` on the list `need_check`. 

Exemple 2: 

- Input: price_split = `[[234],[895]]` The last element of the value in price split is `[895]`. The length of the last element is 3, as it has 3 characters. Then we will append a `True` on the list `need_check`.

In [65]:
need_check = []
len_split = []
for val in products['price_split']:    
    len_split.append(len(val))
    if len(val[-1]) > 2: 
        need_check.append(True)
    else:
        need_check.append(False)

We have created two lists: 

* `need_check`: tells us which element of the dataframe has a last element on `price_split` with a lenght higher than 2. 
* `len_split` : tells us how many elements are inside each `price_split`. 
    * Example 1: `[[59],[99]]` it has a lenght of 2
    * Example 2: `[59]` it has a lenght of 1
    * Example 3: `[[654],[998]]` it has a lenght of 2
    * Example 4: `[[9],[654],[998]]` it has a lenght of 3
    
Then we can add this elements into our products dataframe to detect these rows: 

In [66]:
(
products
    .assign(need_check = need_check, 
            len_split = len_split)
    .query('need_check==True & len_split > 1 & dots_count != 0')
).head()

sku                                               name  \
34   TWS0019  Twelve South MagicWand support Apple Magic Tra...   
362  REP0043                      Speaker lower repair iPhone 4   
480  PIE0011                     Internal Battery for iPhone 3G   
515  SEN0061  Sennheiser EZX 80 Handsfree iPhone iPad and iP...   
518  SEV0026  Service installation RAM + HDD + SSD MacBook /...   

                                                  desc    price promo_price  \
34   MagicWand for wireless keyboard and Magic Trac...  000.000     299.899   
362  Repair service including parts and labor for i...  499.004     499.004   
480        Replacement AC Adapter for Apple iPhone 3G.   98.978      98.978   
515  IPhone bluetooth headset with microphone iPad ...  649.891     649.891   
518  RAM + HDD installation + SSD in your MacBook /...  599.918     599.918   

     in_stock      type price_split  dots_count  need_check  len_split  
34          0      8696  [000, 000]           1        True          2  
362         0  1,44E+11  [499, 004]           1        True          2  
480         0  21485407   [98, 978]           1        True          2  
515         0      5384  [649, 891]           1        True          2  
518         0  20642062  [599, 918]           1        True          2

By defining a set of rules, I can quickly detect the prices that are giving me errors. Those rules are: 

* The length of the last element on the value `price_split` has a lenght bigger than 2 (This is what the column `need_check` is telling us).
* There is more than 1 element inside the column `price_split` (Information given by the column `len_split`). 
* The number of dots in price is different than 0 (Column `dots_count`). 

The next steps would be to use these conditions to replace this values by missing values. To do it, I am going to use the function from the module `numpy` called `np.where`. It works as an `if else` statement; if a condition is accomplished, then return a value, else, return another. 

In [67]:
products = (
products
    .assign(need_check = need_check, 
            len_split = len_split)
    .assign(
        price = lambda x: pd.to_numeric(
            np.where((x['need_check']==True) & (x['len_split'] > 1) & (x['dots_count'] != 0), np.nan, x['price'])
        )
    )
)
products.query('need_check==True & len_split > 1 & dots_count != 0').head(10)

sku                                               name  \
34   TWS0019  Twelve South MagicWand support Apple Magic Tra...   
362  REP0043                      Speaker lower repair iPhone 4   
480  PIE0011                     Internal Battery for iPhone 3G   
515  SEN0061  Sennheiser EZX 80 Handsfree iPhone iPad and iP...   
518  SEV0026  Service installation RAM + HDD + SSD MacBook /...   
525  SEV0024      Service installation RAM + HDD + SSD Mac mini   
526  SEV0025          Service installation RAM + HDD + SSD iMac   
528  SEV0027            Budget Request repair Apple MacBook Air   
532  PAC0362                NewerTech pack 2 cups 225 "for iMac   
536  SEV0028            Budget Request repair Apple MacBook Pro   

                                                  desc  price promo_price  \
34   MagicWand for wireless keyboard and Magic Trac...    NaN     299.899   
362  Repair service including parts and labor for i...    NaN     499.004   
480        Replacement AC Adapter for Apple iPhone 3G.    NaN      98.978   
515  IPhone bluetooth headset with microphone iPad ...    NaN     649.891   
518  RAM + HDD installation + SSD in your MacBook /...    NaN     599.918   
525  installation RAM HDD + SSD + on your Mac mini ...    NaN     599.918   
526  installation RAM HDD + SSD + on your iMac + Da...    NaN     599.918   
528                   diagnosis to repair MacBook Air.    NaN     199.892   
532     Two suction pads 225 inches for aluminum iMac.    NaN     169.896   
536                   diagnosis to repair MacBook Pro.    NaN     199.892   

     in_stock      type price_split  dots_count  need_check  len_split  
34          0      8696  [000, 000]           1        True          2  
362         0  1,44E+11  [499, 004]           1        True          2  
480         0  21485407   [98, 978]           1        True          2  
515         0      5384  [649, 891]           1        True          2  
518         0  20642062  [599, 918]           1        True          2  
525         0  20642062  [599, 918]           1        True          2  
526         0  20642062  [599, 918]           1        True          2  
528         0  1,46E+11  [199, 892]           1        True          2  
532         1  14305406  [199, 795]           1        True          2  
536         0  1,46E+11  [199, 892]           1        True          2

As you can see, we replaced all the values with wrong prices with missing data. New let's drop all the columns we do not need: 

In [68]:
products.drop(['price_split','need_check','len_split','dots_count'], axis=1, inplace=True)
check_missing_values(products)

Missing values:
sku              0
name             0
desc             7
price          587
promo_price      0
in_stock         0
type            50
dtype: int64


And now is time to find a strategy to fill this missing values. 

In [69]:
products.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10579 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sku          10579 non-null  object 
 1   name         10579 non-null  object 
 2   desc         10572 non-null  object 
 3   price        9992 non-null   float64
 4   promo_price  10579 non-null  object 
 5   in_stock     10579 non-null  int64  
 6   type         10529 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 661.2+ KB


#### Missing values

As now we have incremented the number of missing data (we transformed all the wrong prices to missing ones) let's find out an strategy to fill them. 

But first, let's look if they are important in our dataset:

In [70]:
products.assign(row_na = products.isna().sum(axis=1) > 0).query('row_na == False')

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19321  BEL0376        Belkin Travel Support Apple Watch Black   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc  price promo_price  \
0           Aluminum support compatible with all MacBook  59.99     499.899   
1              USB ultrathin keyboard Apple Mac Spanish.  59.00     589.996   
2                                 mouse Apple USB cable.  59.00     569.898   
3                  IPhone dock and USB Cable Apple iPod.  25.00     229.997   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
...                                                  ...    ...         ...   
19321  compact and portable stand vertically or horiz...  29.99     269.903   
19322  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19323  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19324  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19325  Backpack with capacity of 14 liter compartment...  69.95     649.903   

       in_stock      type  row_na  
0             1      8696   False  
1             0  13855401   False  
2             0      1387   False  
3             0      1230   False  
4             1      1364   False  
...         ...       ...     ...  
19321         1     12282   False  
19322         1      1392   False  
19323         1      1392   False  
19324         0      1392   False  
19325         1      1392   False  

[9940 rows x 8 columns]

In [71]:
prod_na = products.loc[products.isna().sum(axis=1) > 0,:].copy()
prod_na

sku                                               name  \
34     TWS0019  Twelve South MagicWand support Apple Magic Tra...   
307    SAN0017       SanDisk Cruzer Edge USB 2.0 Flash Drive 16GB   
362    REP0043                      Speaker lower repair iPhone 4   
480    PIE0011                     Internal Battery for iPhone 3G   
515    SEN0061  Sennheiser EZX 80 Handsfree iPhone iPad and iP...   
...        ...                                                ...   
19312  REP0424                       Input repair Headphones iPad   
19313  REP0421                     iPad charging connector repair   
19314  REP0416                           iPad front camera repair   
19315  REP0413                            repair rear camera iPad   
19316  REP0407                                iPad battery repair   

                                                    desc  price promo_price  \
34     MagicWand for wireless keyboard and Magic Trac...    NaN     299.899   
307                         Pendrive USB 2.0 Mac and PC.    8.0      59.895   
362    Repair service including parts and labor for i...    NaN     499.004   
480          Replacement AC Adapter for Apple iPhone 3G.    NaN      98.978   
515    IPhone bluetooth headset with microphone iPad ...    NaN     649.891   
...                                                  ...    ...         ...   
19312  Repair service including parts and labor for iPad    NaN       69.99   
19313  Repair service including parts and labor for iPad    NaN       69.99   
19314  Repair service including parts and labor for iPad    NaN       69.99   
19315  Repair service including parts and labor for iPad    NaN       69.99   
19316  Repair service including parts and labor for iPad    NaN       69.99   

       in_stock      type  
34            0      8696  
307           0       NaN  
362           0  1,44E+11  
480           0  21485407  
515           0      5384  
...         ...       ...  
19312         0  1,44E+11  
19313         0  1,44E+11  
19314         0  1,44E+11  
19315         0  1,44E+11  
19316         0  1,44E+11  

[639 rows x 7 columns]

Decide which percentage of the data represents. Always think smart about using your time: 

* Which % of the orders contains one of these products? In that case I has to look at orderlines.

In [72]:
orderlines.shape[0]

293743

In [73]:
# percentage of orders with products with missing prices on products dataset
orderlines.loc[orderlines['sku'].isin(prod_na['sku']),:].shape[0] / orderlines.shape[0]

0.025604014393534484

In [74]:
# this is interesting, as for the products data the percentage is much smaller
prod_na.shape[0] / products.shape[0]

0.06040268456375839

Which impact has on the total revenue?

In [75]:
orderlines.loc[orderlines['sku'].isin(prod_na['sku']),:]['total_price'].sum() / orderlines['total_price'].sum()

0.027227081578374515

With a 3% only on the total revenue, but we are considering all the orders, and we are interested on the completed ones. 

In [76]:
na_skus = prod_na['sku'].tolist()

(
orderlines
    .merge(orders, how='left', left_on='id_order', right_on='order_id')
    .query('state == "Completed" & sku == @na_skus')
    ['total_price'].sum()
) / (
orderlines
    .merge(orders, how='left', left_on='id_order', right_on='order_id')
    .query('state == "Completed"')
    ['total_price'].sum()
)

0.02922734355997912

As you can see, the products that are missing only represents 2.9% of the total invoice. 
I will not drop them, as it still has some information which I would like to keep it to create categories (sku, item name and item description), so I am going to save it with prices as missing values. But I will add an extra column telling me which is the maximum price for these skus in orderlines.

In [77]:
# find out which sku are missing in products
prod_na_list = (
products
    .loc[products['price'].isna(),'sku']
    .tolist()
)
prod_na_list

# find them on orderlines, and aggregate the information to find the maximum value
orderlines_prod_na_max = (
orderlines
    .query("sku == @prod_na_list")
    .groupby('sku') 
    .agg({'unit_price':'max'})
    .rename(columns={'unit_price':'max_price_orderlines'})
    .reset_index()
)
orderlines_prod_na_max[:5]

sku  max_price_orderlines
0  ADN0023-A                 16.81
1  APP0660-A                253.11
2    APP0869                539.00
3    APP0870                539.00
4    APP0871                539.00

Time to add the column `max_price_orderlines` in our products data:

In [78]:
products = (
products
    .merge(orderlines_prod_na_max, how='left',on='sku')
)
products.loc[products['max_price_orderlines'].notna()].sample(10)

sku                                               name  \
4311    CEL0051   Celly 4000mAh Battery Power Bank Aluminum Dorado   
5137  CRU0044-A                Open - Crucial MX300 275GB SSD Disk   
7544    PAC1986  Synology DS1817 + | 16GB RAM | 24TB (8x3TB) WD...   
4974    SAN0139            SanDisk SSD 480GB Plus 25 "SATA 6Gb / s   
1262    PAC0562  Kingston V300 SSD expansion kit for Mac mini 1...   
6478    PAC1906  SSD Crucial MX300 expansion kit 1TB + 16GB 160...   
8338  IFX0049-A  Open - iFixit battery Macbook Pro 13 "OEM (Mid...   
7524    PAC1966                        Synology DS1817 + | 8GB RAM   
3656    DLK0128               D-Link DES-1005P PoE Ethernet Switch   
6878    REP0264  volume button repair on vibration control and ...   

                                                   desc  price promo_price  \
4311  external battery capacity 4000mAh output volta...    NaN     129.894   
5137  SSD 275GB SATA Hard Disk 3.0 (6Gb / s) for Mac...    NaN     918.726   
7544  NAS server with 16GB of RAM and 24TB for Mac a...    NaN  19.537.459   
4974                           Hard SSD 480GB 25 inches    NaN   1.565.837   
1262  SSD upgrade kit for Mac mini 120GB 2011 Late 2...    NaN   1.065.841   
6478  SSD expansion kit 1TB + 16GB 1600Mhz RAM for M...    NaN   4.515.841   
8338       Reconditioned OEM battery for MacBook Pro 13    NaN      80.423   
7524  NAS with 8GB RAM and 8 bays compatible with SA...    NaN  10.719.898   
3656            compact switch with 5 Ethernet (PoE 1).    NaN     589.899   
6878  Repair service including parts and labor for i...    NaN     599.906   

      in_stock      type  max_price_orderlines  
4311         0      1515                 24.99  
5137         0      1298                 77.86  
7544         0  12175397               2055.75  
4974         0  12215397                156.58  
1262         0      1433                 89.99  
6478         1      1433                483.59  
8338         0      1298                 80.22  
7524         0  12175397               1099.99  
3656         0      1334                 46.90  
6878         0  1,44E+11                 59.99

And now let's store it as a products clean csv into our path: 

In [79]:
products.drop(['promo_price','in_stock','type'], axis=1).to_csv(path + 'products_cl.csv', index=False)

## Brands

The brands csv looks fine, so we can work together with it. 

In [80]:
brands.to_csv(path + 'brands_cl.csv', index=False)